Add new Stripe gateways UI

cesarcosta99 opened the pull request:
# Changes proposed in this Pull Request:

Resolves #1704 

This PR implements the new UI for the Stripe gateways outside of UPE. I made it reusable so that we can easily add all the gateways we have and others as they come in.

I've chose to build the new UI inside the UPE settings scope. Some reasons for that are: a) it can be considered part of the settings, even though it's not relying on UPE; b) the UI is very similar to what we have in UPE settings so we could reuse many of the components; c) I think it's not worth to create a new build/entrypoint to something that small, given its context as part of the settings and its reusability.

In the backend, there's a tradeoff I'm considering too. While taking advantage of the settings REST controller we already have, we'll be adding some new fields to the REST controller and that will bring more complexity and amplify its scope. However, the other option would be to create a new REST controller to handle the other gateways and I presume most (maybe all) of them will be removed soon, when UPE becomes mandatory, and this new controller will become useless. Besides, creating a new controller would force us to create a whole new structure in the React-side as well, which would become useless at some point either.

# Testing instructions
Ideally you must disable the new experience checkout so that you can test all the payment gateways, including those that come built-in with UPE.

- Ensure you have the option `_wcstripe_feature_upe_settings` set to `yes` in the database.
- Visit the [WooCommerce Payment settings page](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout).
- Access any of the Stripe gateways settings, such as SEPA, Alipay, etc
- They should look similar to the screenshot below.
- You should be able to enable/disable and edit their name and description as you would like.
- Repeat the previous step and try to edit each payment gateway to confirm they're all working: SEPA, giropay, iDeal, Bancontact, EPS, SOFORT, P24, Alipay and Multibanco.

![Screen Shot 2021-08-19 at 9 00 33 AM](https://user-images.githubusercontent.com/273592/130082071-9424bbd7-3038-4ced-951f-2e22c9299268.png)

# What's still missing
- [x] Create issue to handle the webhook monitoring date

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

cesarcosta99 committed:
Update content of payment gateways' settings

Only applicable when the UPE settings redesign is enabled.

cesarcosta99 committed:
Add new payment gateway screen

cesarcosta99 committed:
Add UI to the new gateway screen

cesarcosta99 committed:
Include settings enqueue in stripe gateways

cesarcosta99 committed:
Add gateway descriptions

cesarcosta99 committed:
Add hooks for alipay and multibanco settings

cesarcosta99 committed:
Add get gateway helper

cesarcosta99 committed:
Add payment gateway section hooks

cesarcosta99 committed:
Connect payment gateway section data

cesarcosta99 committed:
Add tests for the payment gateway screen

cesarcosta99 commented:
@frosso I wonder if you have any thoughts/concerns to share with this approach.

cesarcosta99 committed:
Fix test_admin_options_when_stripe_is_connected test

cesarcosta99 committed:
Add tests for the payment gateway section

cesarcosta99 committed:
Fix UPE opt-in banner positioning in payment gateways settings

cesarcosta99 committed:
Inlcude the other payment gateways in the settings enqueue scripts

cesarcosta99 committed:
Add the new UI to SEPA

cesarcosta99 committed:
Add i18n to payment gateways titles

cesarcosta99 committed:
Add the new UI to giropay

cesarcosta99 committed:
Add the new UI to iDeal

cesarcosta99 committed:
Fix naming in the checkbox labels in the payment gateway screens

cesarcosta99 committed:
Rename gateways descriptions to gateways info

cesarcosta99 committed:
Add the new UI to Bancontact

cesarcosta99 committed:
Refactor getGateway to make it dynamic

cesarcosta99 committed:
Add the new UI to EPS

cesarcosta99 committed:
Add the new UI to SOFORT

cesarcosta99 committed:
Add the new UI to P24

cesarcosta99 committed:
Add helper methods to payment gateway around enabled option

cesarcosta99 committed:
Add payment gateway settings logic to rest controller

cesarcosta99 committed:
Add tests for payment gateway section helpers

bborman22 committed:
Merge branch 'develop' into add/1704-new-gateways-ui

frosso commented on client/settings/payment-gateway-section/index.js:
```diff
@@ -0,0 +1,114 @@
+							.toISOString()
+							.replace( 'T', ' ' )
+							.replace( /:\d{2}\..*/, ' UTC' )
+					) }
```
Is this accurate?

cesarcosta99 replied:
No, however I was planning to tackle that in another PR, created #2082.

frosso replied:
Gotcha! FWYW, there's `useAccount().data.webhook_status_message` that will return the correct messaging 👍

frosso commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -24,20 +24,184 @@ class WC_REST_Stripe_Settings_Controller extends WC_Stripe_REST_Base_Controller
+	 * @return WC_Gateway_Stripe|null WC_Gateway_Stripe instance
+	 */
+	private function instantiate_gateway_settings() {
+		$referer         = isset( $_SERVER['HTTP_REFERER'] ) ? esc_url_raw( wp_unslash( $_SERVER['HTTP_REFERER'] ) ) : '';
```
What do you think about having a separate controller for these settings, instead?
I'm not sure using the `HTTP_REFERER` would be "safe", since not all browsers send it to the server.

The new controller could be a dynamic route that explicitly contains the gateway's name.
This way we can also have separate hooks (and maybe even a separate data store) for the `PaymentGatewaySection` component. This way we can also avoid having separate hooks in the `client/data/settings/hooks.js` file, which are not needed by the other settings pages.

cesarcosta99 replied:
Did you get a chance to read my trade-off decision in the description? In short, I also agree that architecture-wise, it's a good idea to move to a new controller and data store. On the other hand, I think investing in a structure that is very likely to become unused soon (once all gateways are moving towards UPE and UPE will be mandatory) might not be the best approach. 

With regards to the `HTTP_REFERER`, I think the [browser compatibility](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer#browser_compatibility) is good enough to cover the downside of using it.

frosso replied:
It is true that most of the gateways will be embedded in the UPE settings UI, but not all of them will.
For example, Multibanco will not be part of UPE for the foreseeable future 🤷 

Another benefit of having a separate controller/storage is that it will probably be easier to detangle everything when cleaning up the settings when UPE will be enabled by default.

Not a blocker, tho - I understand that it would be quite the refactor from this approach.

cesarcosta99 replied:
Good point 👍 I've decided to step back and switch to that approach. I have started the refactoring in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2068#commits-pushed-884509b

bborman22 reviewed (commented):
This all worked well for me overall. The new settings pages rendered correctly for me and once I worked out the issue I'll describe below, they were updating the settings correctly for me.

The one issue I ran in to that might just be an edge case, but I think is relevant here due to the approach taken (which I generally agree with). The issue I encountered was when trying to save any of the settings it was failing with no actionable error message. It turned out that the `short_statement_descriptor` field was empty and thus failing validation. This was unexpected as I didn't have the `is_short_statement_descriptor_enabled` option checked. This validation bug is unrelated to this PR and I'll create an issue to confirm its functionality.  However, it came up in this PR as I was unable to save any of the other gateway settings due to this validation error and also was unable to address the error since it was from a field not present on any gateway settings page. 

To reproduce this if a short statement description has already been provided, the stripe settings could be modified directly in the database to be an empty string. At that point the error should be encountered on any attempt to save settings including on the gateway settings pages.

I don't think this is a showstopper of an issue especially since the original validation error is unrelated to this PR, but I wanted to raise it here to see if there was anything we. could do to smooth over that user experience. In short, if there is some validation that is failing from the main settings page, it seems to prevent saving any gateway level settings as well. Is there anything we can do to make that user experience a bit better? Just leaving as a comment here to see if there is a strong feeling on how best to handle this (as part of this PR or another issue).

![Giropay-settings-failed](https://user-images.githubusercontent.com/68524302/137937002-1b700aaf-92b6-44e1-b703-ef677d00f3b2.png)

![Giropay-settings-failed-reason](https://user-images.githubusercontent.com/68524302/137937046-87d51621-ab29-42ee-9b0d-728abb4089f8.png)

frosso reviewed (commented):
Thanks for the ping @cesarcosta99 and apologies for the delayed review, I got to finish it only now!

cesarcosta99 commented:
@bborman22 thanks a bunch for reviewing and for bring that issue up. I'm so sorry for not mentioning it before, but it's indeed something that is affecting the expected behavior on saving it. This issue is reported under #2047 and I've already addressed it on #2074.

FWIW, it should work if you enable `Add customer order number to the bank statement`, provide a 5~10-length `Shortened customer bank statement` and save.

>I don't think this is a showstopper of an issue especially since the original validation error is unrelated to this PR, but I wanted to raise it here to see if there was anything we. could do to smooth over that user experience. In short, if there is some validation that is failing from the main settings page, it seems to prevent saving any gateway level settings as well. Is there anything we can do to make that user experience a bit better? Just leaving as a comment here to see if there is a strong feeling on how best to handle this (as part of this PR or another issue).

This is a very special condition met that was not caused due to data provided but instead, due to a bug introduced of a wrong validation rule. With that said, if data is saved and validation rules are correct, I think it should never prevent the gateway settings from being saved. One thing we could do to be extra careful, though, would be to prevent validation from happening in the general settings fields when `$this->gateway_settings` is set.

cesarcosta99 committed:
Restrict settings update operation

Since the approach taken is using the settings REST controller to update both the actual settings and the specific gateways settings, we're restricting the update scope of the gateway settings so that the validation of the settings fields doesn't interfere in the gateway settings update.

cesarcosta99 committed:
Update webhook messaging on new gateway settings

cesarcosta99 committed:
Update payment gateway section tests

cesarcosta99 commented:
>One thing we could do to be extra careful, though, would be to prevent validation from happening in the general settings fields when $this->gateway_settings is set.

Addressed in bf8f184.

Since the newly introduced webhook messaging hook is already here, I added it in 2f823ef.

frosso committed:
Merge branch 'develop' into add/1704-new-gateways-ui

frosso commented on client/settings/payment-gateway-manager/index.js:
```diff
@@ -0,0 +1,59 @@
+				</LoadableSettingsSection>
+			</SettingsSection>
+			<SettingsSection>
+				<UpeOptInBanner />
```
It looks like there's some inconsistency with the spacing between the designs and the UI:
Designs: https://d.pr/i/qQtIcq
This branch: https://d.pr/i/uIaDEF

The banner that needs to be displayed on the individual payment methods settings needs to have a different background image: https://d.pr/i/Xd4rn9 (see `develop` branch)

cesarcosta99 replied:
That's true, the text is different too, thanks for spotting that! Updated it in f7a2005 and 6f77c49.

frosso commented on tests/phpunit/test-wc-stripe-payment-gateway.php:
```diff
@@ -52,7 +52,7 @@ public function test_admin_options_when_stripe_is_connected() {
 		ob_start();
 		$this->giropay_gateway->admin_options();
 		$output = ob_get_clean();
-		$this->assertStringMatchesFormat( '%aid="wc-stripe-upe-opt-in-banner"%a', $output );
```
Since this has been removed, then also the scripts should be removed: `client/entrypoints/upe-opt-in-banner/index.js` and the enqueue for `woocommerce_stripe_upe_opt_in` (as well as the webpack config)

cesarcosta99 replied:
Good catch, addressed in 37baf34 and 4376482 👍

frosso commented on client/data/settings/hooks.js:
```diff
@@ -108,6 +108,105 @@ export const useGetAvailablePaymentMethodIds = makeReadOnlySettingsHook(
 );
 
+// SEPA
+export const useIsStripeSepaEnabled = makeSettingsHook(
```
Rather than manually creating all these hooks, what do you think about some generic ones by leveraging the currying pattern?

```
export const makeIsPaymentMethodEnabledHook = ( methodName ) =>
	makeSettingsHook( `is_${ methodName }_enabled` );

export const makePaymentMethodNameHook = ( methodName ) =>
	makeSettingsHook( `${ methodName }_name`, '' );

export const makePaymentMethodDescriptionHook = ( methodName ) =>
	makeSettingsHook( `${ methodName }_description`, '' );
```

Then `client/settings/payment-gateway-section/hooks.js` can just import `makeIsPaymentMethodEnabledHook`, `makePaymentMethodNameHook` and `makePaymentMethodDescriptionHook` (without importing all the hooks).

This should help a little bit with the bundle size (since we wouldn't have all these hooks) and maybe even a tiny bit with tree shaking (since we'd be explicitly declaring what is being imported, instead of leveraging `*`).

Then, instead of having `getGateway()` we can just leverage `const { section } = getQuery();` in `client/settings/payment-gateway-section/hooks.js` to get the current payment method name.

cesarcosta99 replied:
Nice, I haven't thought about this! Applied it in 1a20156.

cesarcosta99 committed:
Create payment gateway settings rest controller

cesarcosta99 committed:
Import and instantiate the payment gateway rest controller

cesarcosta99 committed:
data(payment gateway): add action types

cesarcosta99 committed:
data(payment gateway): add reducer

cesarcosta99 committed:
data(payment gateway): add actions

cesarcosta99 committed:
data(payment gateway): add resolvers

cesarcosta99 committed:
data(payment gateway): add selectors

cesarcosta99 committed:
data(payment gateway): add hooks

cesarcosta99 committed:
data(payment gateway): set up index module

cesarcosta99 committed:
data: include payment gateway in the store

cesarcosta99 committed:
Create save payment gateway section

cesarcosta99 committed:
Update save section in payment gateway manager

cesarcosta99 committed:
Update hooks in payment gateway section

cesarcosta99 committed:
Remove code and files from previous approach

This is part of a refactor and there was a change in the approach, please refer to #2068 for more context.

frosso commented on client/data/payment-gateway/actions.js:
```diff
@@ -0,0 +1,58 @@
+import { NAMESPACE, STORE_NAME } from '../constants';
+import ACTION_TYPES from './action-types';
+
+const { section } = getQuery();
```
(minor) technically this could fail if we were to implement soft routing between pages - `section` would be instantiated only once, when the file is imported.
An alternative could be to move it within the action where it's used (the same is also true for `client/data/payment-gateway/hooks.js`)

cesarcosta99 replied:
Good point, fixed it in 79b61e4 and 4d03746. I had to make the hooks implementation a little bit different because the hooks are instantiated when the file is imported, if I refactor it to functions I'd have to make an [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE).

frosso commented on client/data/payment-gateway/hooks.js:
```diff
@@ -0,0 +1,87 @@
+
+export const usePaymentGatewayDescription = makePaymentGatewayDescriptionHook(
+	section
+);
```
What do you think about just inlining some of these, since they're not exported anywhere else?

```suggestion
export const useEnabledPaymentGateway = makePaymentGatewayHook( `is_${ section }_enabled` );

export const usePaymentGatewayName = makePaymentGatewayHook( `${ section }_name`, '' );;

export const usePaymentGatewayDescription = makePaymentGatewayHook( `${ section }_description`, '' )
```

cesarcosta99 committed:
Remove enqueue for admin_scripts_for_banner

cesarcosta99 committed:
Remove upe-opt-in-banner entrypoint

cesarcosta99 committed:
Add payment gateway banner image

cesarcosta99 committed:
Update payment gateway opt-in banner

cesarcosta99 committed:
Update payment gateway loadable component

cesarcosta99 committed:
Fix payment gateway section tests

cesarcosta99 committed:
data(payment gateway): Add actions tests

cesarcosta99 committed:
data(payment gateway): Add hooks tests

cesarcosta99 committed:
data(payment gateway): Add reducer tests

cesarcosta99 committed:
data(payment gateway): Add resolvers tests

cesarcosta99 committed:
data(payment gateway): Add selectors tests

cesarcosta99 committed:
Add sanitization to text fields

cesarcosta99 committed:
Add boolean validation to field on update

cesarcosta99 committed:
Add tests for payment gateway REST controller

cesarcosta99 committed:
Fix lint error

frosso commented on client/data/payment-gateway/resolvers.js:
```diff
@@ -0,0 +1,27 @@
+ * Retrieve payment gateway settings from the site's REST API.
+ */
+export function* getPaymentGateway() {
+	const path = `${ NAMESPACE }/payment-gateway-settings/${ section }`;
```
(minor) What are your thoughts on aligning the naming between the functions, the Redux store name and the endpoint?
i.e.: the endpoint is called `payment-gateway-settings/*`, but the store name is `paymentGateway`, the container name is `wc-stripe-payment-gateway-container` and the messages displayed to the user refer to the payment gateway's settings. Can we maybe name the store (and hooks) `paymentGatewaySettings`? Or would that be too verbose?

cesarcosta99 replied:
I think that's too verbose but you indeed have a good point. I decided to rename the endpoint to `payment-gateway/*` in 17b93b4. I also updated the file names to make it more consistent.

frosso committed:
Merge branch 'develop' into add/1704-new-gateways-ui

frosso commented on client/settings/payment-gateway-manager/opt-in-banner.js:
```diff
@@ -0,0 +1,36 @@
+			data-testid="opt-in-banner"
+		/>
+	);
+};
```
In the previous settings we did not display the banner on Multibanco and Alipay, since they are not yet available in UPE.
![Screen Shot 2021-10-27 at 11 26 55 AM](https://user-images.githubusercontent.com/273592/139107400-5f1d512e-6ecc-459a-9480-8f89f3382194.png)
![Screen Shot 2021-10-27 at 11 26 59 AM](https://user-images.githubusercontent.com/273592/139107382-d9d6b173-dafd-472f-b500-410008865bf0.png)


However, in this branch I'm seeing the banner on those payment methods:
![Screen Shot 2021-10-27 at 11 27 55 AM](https://user-images.githubusercontent.com/273592/139107542-2050a25e-a601-4dbb-a94d-8786580a6a74.png)

![Screen Shot 2021-10-27 at 11 27 44 AM](https://user-images.githubusercontent.com/273592/139107507-d80fe11c-25b1-4bdf-a732-23050c0309a4.png)

cesarcosta99 replied:
Good catch, I completely missed it. Addressed in 9b55aa5.

frosso reviewed (approved):
I tested the functionality, this is looking really good!
It would be ideal to remove the UPE banner from the payment methods that cannot be used with the UPE before merging.

Now it's just a matter of resolving the tests :) 
Thanks for indulging my comments!

cesarcosta99 committed:
Update payment gateway hooks

Co-authored-by: Francesco <frosso@users.noreply.github.com>

cesarcosta99 committed:
Hide opt-in banner on some gateways

Hiding it basically on payment gateways that are not available on UPE yet.

cesarcosta99 committed:
Fix payment gateway action implementation

cesarcosta99 committed:
Fix payment gateway hooks implementation

cesarcosta99 committed:
Renamed payment gateway settings class and endpoint

cesarcosta99 committed:
Renamed payment gateway settings endpoint in FE

cesarcosta99 committed:
Fix missing return statement

cesarcosta99 committed:
Fix payment gateway hooks implementation

cesarcosta99 committed:
Add tests for save payment gateway component

cesarcosta99 committed:
Fix payment gateway REST controller tests

cesarcosta99 committed:
Merge branch 'develop' into add/1704-new-gateways-ui
